### ISSUE TITLE
Inheritance Support?

### ISSUE THREAD

**Issue by ahmedk92** [editfmah/sharkorm#7]
Although not mentioned in the documentation, I tried subclassing a custom entity I made. But I get the following exception `'NSInvalidArgumentException', reason: '*** setObjectForKey: key cannot be nil'`

If direct subclassing is not supported, is there any alternative?

**Comment by editfmah** [comment]
I'll just check now, it should work fine. Although i've never tested that in swift.

**Comment by editfmah** [comment]
Looks like you can't subclass SRKObjects, this seems to be the same with some of the other ORM's too. I checked them out to see how they solved the introspection issue, and they don't by the looks of it.

Realistically, the only "work around" would be to embed one object within the other, but it will most likely cause other issues with your queries.

So:

``` swift
class Person : SRKObject {
    internal dynamic var Name: String?
    internal dynamic var age: NSNumber?
    internal dynamic var department: DepartmentSwift?
}

class SmallPerson : Person {
    dynamic var height: NSNumber?
}
```

Would become:

``` swift
class SmallPerson : SRKObject {
    dynamic var person: Person?
    dynamic var height: NSNumber?
}
```

Then you would have to assign a new `Person` object when creating a `SmallPerson`, but this would litter the `Person` table.  So you would have to exclude the `SmallPerson` related objects from and queries, and this would be slow and cumbersome. 

I will add it to the features list, as it would be a nice distinction and advantage over other ORMs.

We are quite excited by this potential though, so we will have a quick play with a prototype now.  Basically we inspect the current class for information to build the tables, but we could also look at the superclass too, maybe?

**Comment by ahmedk92** [comment]
I think Realm Swift say they support inheritance. But they recommend the composition pattern as you described here. But composition introduce other problems in data mapping (JSON serialization for example), as you cannot have control over the API.

**Comment by editfmah** [comment]
Yeah, it was realm I looked at. They say they do, but the SO articles from users cloud that picture somewhat.  I think it does something odd with the table structure, which we would seek to avoid.  We are just cranking some code in now, it should be a pretty simple change.  But if it does work, it might take a week or so to fully test all angles to ensure nothing is broken before we could release.

**Comment by editfmah** [comment]
Totally works, thank you so much for suggesting it.  Turned out to be very simple in the end with minor changes.  

We inject and alter schemas from a dictionary of properties and their types, so just looping up the class chain solved this issue.  All the test passed, and if you like i'll branch this code and you can try it for yourself.  Chances are, considering the very small nature of the change, that it will be tested in the next day or so.

**Comment by editfmah** [comment]
Here is the branch should you wish to try it out.

https://github.com/sharksync/sharkorm/tree/subclass_support

**Comment by editfmah** [comment]
This has now been released as version 2.0.4.

**Comment by alldritt** [comment]
I'm experimenting with the subclass support and I have a query question.  Say I have Person and SmallPerson objects; is there a way to query both Person and SmallPerson tables and get back a collection of all the Person *and* SmallPerson objects satisfying the query?

I tried adding this to the subclass tests:

```
- (void)test_person_object {
    [self setupCommonData];
    
    SRKResultSet *r = [[Person query] where:@"age = 37"].fetch;
    
    XCTAssert(r,@"Failed to return a result set");
    XCTAssert(r.count == 1,@"incorrect number of results returned");
    
    Person* p = [r firstObject];
    
    NSLog(@"p %@", p);
}
```

**Comment by Parkyprg** [comment]
@alldritt - say you have this structure:

###SRKObject SmallPerson###
@property NSString *name;

###SRKObject Person###
@property NSInteger age;
@property SmallPerson *smallPerson;

If you query for persons, then the resulted objects will also have a link to the related small person:
SRKResultSet *res = [[Person query] where:@"age = 37"].fetch;

You can access res.smallPerson.name right away.

**Comment by editfmah** [comment]
@alldritt ,

From an internal point of view it would be complicated to achieve because a "query" has a base table and not several, but you could also have duplicate primary key values in the two tables which would also result in some fairly complicated workarounds.

Much better solution to would to execute two simultaneous queries and then combine the results?

```
typedef id(^ExecutionBlock)();
typedef void(^ExecutionCompletionBlock)(id result1, id result2);

+ (void)execute_parallel_with_result:(ExecutionBlock)closure1 closure2:(ExecutionBlock)closure2 completion:(ExecutionCompletionBlock)completion {
	
	id var_lock = [NSObject new];
	NSLock* proc_lock = [NSLock new];
	__block int outstandingClosures = 0;
	__block id result1 = nil;
	__block id result2 = nil;
	
	if(closure1){outstandingClosures++;};
	if(closure2){outstandingClosures++;};
	
	if (outstandingClosures) {
		
		[proc_lock lock];
		
		if (closure1) {
			[ExecutionHelper execute_background:^{
				result1 = closure1();
				@synchronized(var_lock) {
					outstandingClosures--;
					if (!outstandingClosures) {
						[proc_lock unlock];
					}
				}
			}];
		}
		
		if (closure2) {
			[ExecutionHelper execute_background:^{
				result2 = closure2();
				@synchronized(var_lock) {
					outstandingClosures--;
					if (!outstandingClosures) {
						[proc_lock unlock];
					}
				}
			}];
		}
		
		[proc_lock lock];
		[proc_lock unlock];
		
	}
	
	completion(result1, result2);
	
} 
```

**Comment by MeteC** [comment]
Just noticed this in the feed - if inheritance works well, this is brilliant! Would love to see this feature in the documentation 😉

**Comment by editfmah** [comment]
@MeteC  
Yeah, inheritance works as well as we know and our testing goes. We certainly use it in our major projects, we just don't allow mixing of class types in a query.

**Comment by alldritt** [comment]
I appreciate the suggestions.  It appears that @Parkyprg's suggestion is the most workable.  However, I simplified my use case to create a simple example.  My real use involves multiple subclasses:

```
class Recording : SRKObject {
    ...
}

class RawRecording : Recording {
    ...
}

class OddballRecording : Recording {
    ...
}
```

The question that arises is how to create the link from a Recording instance to its associated RawRecording or OddballRecording instance.

**Comment by editfmah** [comment]
Okay, so in Psuedo'ish code it would look like this:

**"Version A", where a recording can have both as properties:**

```
class Recording : SRKObject {
    dynamic var raw: RawRecording? = nil
    dynamic var oddball: OddballRecording? = nil
}

class RawRecording : Recording {
    ...
}

class OddballRecording : Recording {
    ...
}

// a usage example being, for say an oddball

let newRecording = Recording()
newRecording.oddball = OddballRecording()
....
newRecording.commit()

// referencing the properties using object dot notation is easy enough.
let oddballLength = newRecording.oddball.length
```

**or Version B, which makes a bit more sense:**

```
class Recording : SRKObject {
    ...
}

class RawRecording : Recording {
    dynamic var recording: Recording? = nil
}

class OddballRecording : Recording {
    dynamic var recording: Recording? = nil
}

// usage
let oddball = OddballRecording()
oddball.recording = Recording()
... //do stuff with object
oddball.commit()

// you would want to create a trigger to clean up the related record upon deletion, to keep it neat.

```

**Comment by alldritt** [comment]
Thank you.  Version B is the way I've chosen to proceed.  Its working fairly well.  The only real frustration is the bookkeeping needed to keep these base/subclass relationships in order at creation and deletion, but its not too onerous.

A thought for the future: it seems like some join magic could be employed to handle polymorphism cleanly.
